pkgs/top-level/release-lib.nix

```diff
@@ -33,7 +34,7 @@ let
     systems
     ;
 
-  pkgs = packageSet (recursiveUpdate { system = "x86_64-linux"; config.allowUnsupportedSystem = true; } nixpkgsArgs);
+  pkgs = packageSet (recursiveUpdate { inherit system; config.allowUnsupportedSystem = true; } nixpkgsArgs);
```

Comment:
This and the other change in this file doesn't seem to be making it purer, but rather makes the system not hardcoded. I guess it wasn't possible to run this on non-x86_64-linux systems before?

Reply:
Before, `builtins.currentSystem` was being used in certain places. This PR passes system around and uses `forAllSystems` so multiple platforms can do the release jobs and such.

Reply:
It is pure changes because we're passing system instead of inferring from the builtins.

Reply:
But the value here didn't come from `builtins.currentSystem`, it was hardcoded before, which is technically also pure. So the change here doesn't make it any more pure, but rather just avoids the hardcoding.

Reply:
Yeah, I just used `inherit system;` since the system would already need to be passed to this file.

Reply:
It looks very sensible, but I think I'm still not understanding a piece of the puzzle: Why was this hardcoded before without ever causing problems? I can't test it locally right now, but I presume `nix flake check` on e.g. Darwin fails if this is left hardcoded? Did nobody on Darwin ever evaluate `release-lib.nix`, or is the `pkgs` it imports never really used perhaps? Perhaps it's hardcoded for a reason, such that you get reproducible results on all systems? Looking at the git blame might be useful

Would be great to figure this out, especially because nobody seems to really know the code here well!

Reply:
Yeah, it was originally written before I joined in so I am not sure why it was never an issue before.

Reply:
At least [this commit](https://github.com/NixOS/nixpkgs/commit/3e9f76774ab03c0aa06267dcfa1c7d7f6c984cf2) gives a bit of context

Reply:
Interesting. I wonder why specifically `x86_64-linux` was chosen as the "base system". Probably because not many people use ARM? I'm on aarch64 so at least this PR makes this work for me.

Reply:
For reference, what failure with which command are you getting on aarch64 before this PR?

Reply:
I would get this:
```
error:
       … while checking flake output 'checks'

         at /nix/store/nryiv97qvy45sm4pxr0kz42rkq74bd42-source/flake.nix:46:7:

           45|
           46|       checks.x86_64-linux = {
             |       ^
           47|         tarball = jobs.tarball;

       … while checking the derivation 'checks.x86_64-linux.tarball'

         at /nix/store/nryiv97qvy45sm4pxr0kz42rkq74bd42-source/flake.nix:47:9:

           46|       checks.x86_64-linux = {
           47|         tarball = jobs.tarball;
             |         ^
           48|         # Test that ensures that the nixosSystem function can accept a lib argument

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: attribute 'currentSystem' missing

       at /nix/store/nryiv97qvy45sm4pxr0kz42rkq74bd42-source/pkgs/top-level/impure.nix:17:43:

           16|   # (build, in GNU Autotools parlance) platform.
           17|   localSystem ? { system = args.system or builtins.currentSystem; }
             |                                           ^
           18|
```

Reply:
No I mean if you leave the hardcoded `x86_64-linux` value here

Reply:
```
error:
       … while checking flake output 'checks'

         at /nix/store/k4bwf1ilnd60pv7vz294ghgrzmazm1am-source/flake.nix:47:7:

           46|
           47|       checks = forAllSystems (system: {
             |       ^
           48|         inherit (jobs.${system}) tarball;

       … while checking the derivation 'checks.x86_64-darwin.nixosSystemAcceptsLib'

         at /nix/store/k4bwf1ilnd60pv7vz294ghgrzmazm1am-source/flake.nix:52:9:

           51|         #       alternatives include: `import` a file, or put a custom library in an option or in `_module.args.<libname>`
           52|         nixosSystemAcceptsLib = (self.lib.nixosSystem {
             |         ^
           53|           pkgs = self.legacyPackages.${system};

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: assertion '(stdenv).isLinux' failed

       at /nix/store/k4bwf1ilnd60pv7vz294ghgrzmazm1am-source/pkgs/os-specific/linux/kernel/generic.nix:87:1:

           86|
           87| assert stdenv.isLinux;
             | ^
           88|
```

Reply:
Oh so you're on `aarch64-darwin` I guess. But `nix flake check --system aarch64-darwin` fails evaluation no matter whether the value here is hardcoded to `x86_64-linux` or not (tested it locally). I'm looking for a command that didn't work previously, but now does because it's not hardcoded anymore.

Reply:
No, I'm on `aarch64-linux`. With `nix flake check` and this PR, the checks pass.

Reply:
But it also makes it where the checks that are in the Flake get executed on Garnix since I use Garnix for CI. Before, I would get the `builtins.currentSystem` missing error and now there is no error.

Reply:
I feel like I'm missing something. I can't reproduce your eval failures as you describe them. [This failure](https://github.com/NixOS/nixpkgs/pull/317772#discussion_r1630569831) I can only get with `--system aarch64-darwin`, but you're saying you're on `aarch64-linux`?

And I'm specifically asking whether removing the _hardcoded_ part here actually helps making any command work. Please try this:

```
❯ nix-env --version
nix-env (Nix) 2.18.2

❯ nix flake check github:ExpidusOS/nixpkgs/fix/flake-check --no-build --system aarch64-linux
warning: unknown flake output 'htmlDocs'
warning: The check omitted these incompatible systems: aarch64-darwin, armv5tel-linux, armv6l-linux, armv7l-linux, i686-linux, mipsel-linux, powerpc64le-linux, riscv64-linux, x86_64-darwin, x86_64-linux
Use '--all-systems' to check all.

❯ nix flake check github:infinisil/nixpkgs/hardcode --no-build --system aarch64-linux
warning: unknown flake output 'htmlDocs'
warning: The check omitted these incompatible systems: aarch64-darwin, armv5tel-linux, armv6l-linux, armv7l-linux, i686-linux, mipsel-linux, powerpc64le-linux, riscv64-linux, x86_64-darwin, x86_64-linux
Use '--all-systems' to check all.
```

The last command uses the hardcoded `x86_64-linux` on top of your branch: https://github.com/infinisil/nixpkgs/commit/1af88e1976e5e08f86b775686c46825515d2d9d3. And evaluation works just fine with it with `--system aarch64-linux`. So unless you noticed something else that was broken, it doesn't seem to be necessary to remove the hardcoding.

Reply:
> And I'm specifically asking whether removing the _hardcoded_ part here actually helps making any command work.

Both commands seem to work until Darwin fails with the `stdenv.isLinux` assert fails. I've pushed a fix to not include the NixOS stuff outside on Darwin. With the hardcoded value, it tries calling a binfmt I have set up for x86. On NixOS on Apple Silicon, getting x86 emulation working is a pain. Making it `inherit system` makes the build work for the Flake check.

> So unless you noticed something else that was broken, it doesn't seem to be necessary to remove the hardcoded value.

Why do we even need hardcoded values in the first place? Imo, it's better to abstract that out to `builtins.currentSystem` or whatever system-specific attribute set something is a part of.

Reply:
Since my system has bintfmt set up, a better look into this issue is on a system without binfmt that is also ARM. Using the hardcode, I get this error:
```
error: a 'x86_64-linux' with features {} is required to build '/nix/store/lmx4147d1l51j1qvwhy7wh3kll5ssyq7-maintainer-teams.json.drv', but I am a 'aarch64-linux' with features {benchmark, big-parallel, gccarch-armv8-a, kvm, nixos-test}
```

Upstream only does:
```
warning: The check omitted these incompatible systems: aarch64-darwin, armv5tel-linux, armv6l-linux, armv7l-linux, i686-linux, mipsel-linux, powerpc64le-linux, riscv64-linux, x86_64-darwin, x86_64-linux
Use '--all-systems' to check all.
```

While with my PR builds the checks.

Reply:
> With the hardcoded value (and removing --all-systems), it tries calling a binfmt I have set up for x86

Ah that's great to know, that's what I'm trying to get at. So the hardcoded value _does_ end up causing problem on Darwin. If it wouldn't have caused problems, removing it would be barking up the wrong tree.

I think this is a separate enough change that it deserves its own commit, with the commit message saying that it fixes building tests on non-x86_64-linux systems.

Reply:
> _does_ end up causing problem on Darwin

Yeah, on any system outside `x86_64-linux`.

Reply:
> I think removing the hardcoded value and the `optionalAttrs` is a separate enough change that it deserves its own commit, with the commit message saying that it fixes building tests on non-x86_64-linux systems.

Would that be too little of changes for their own commits?

Reply:
It definitely doesn't fit into any "make pure" commit, because it's arguably the complete opposite of that, since it removes a hardcoded value and replacing it with something that could be different on everybody's system. So yeah if you don't mind, I think it really deserves its own commit.

Reply:
Done